### PR TITLE
[Hub Generated] Review request for Microsoft.Cache to add version preview/2022-11-01-preview

### PR DESCRIPTION
**Pull request by riyasimon23**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/riyasimon23/azure-rest-api-specs/dev-redisenterprise-Microsoft.Cache-2022-11-01-preview...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1214]: We should be able to refactor this out into its own definition and reference it in both places:
`
"definitions":{

"SkuName": {
          "type": "string",
          "description": "The type of RedisEnterprise cluster to deploy. Possible values: (Enterprise_E10, EnterpriseFlash_F300 etc.)",
          "enum": [
            "Enterprise_E10",
            "Enterprise_E20",
            "Enterprise_E50",
            "Enterprise_E100",
            "EnterpriseFlash_F300",
            "EnterpriseFlash_F700",
            "EnterpriseFlash_F1500"
          ],
          "x-ms-enum": {
            "name": "SkuName",
            "modelAsString": true
          }
}`

and `"name":  {
              "$ref": "#Definitions/SkuName
            }
`(approx)

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:92]: Chan change these to `"$ref": "#Parameters/LocationParameter"`?

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:73]: >locations [](http://example.com/codeflow?start=63&length=9)

Looks like we're missing a declaration of our locations/checkNameAvailability API?

Comment (TimLovellSmith) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:73]: The route is

                routeTemplate: "subscriptions/{subscriptionId}/providers/Microsoft.Cache/locations/{location}/checkNameAvailability",

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1816]: >value [](http://example.com/codeflow?start=9&length=5)

I'd put readonly: True on these

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1812]: >name [](http://example.com/codeflow?start=9&length=4)

I'd put readonly: True on these

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1788]: >name [](http://example.com/codeflow?start=9&length=4)

I'd think we want to put readonly: True on it in the skuDetail context

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1771]: I'd think we want to put readonly: True on these

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1755]: >value [](http://example.com/codeflow?start=9&length=5)

I'd think we want to put readonly: True on these

**Review by TimLovellSmith — CHANGES_REQUESTED** [review]
:clock1:

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (riyasimon23) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:73]: Hi Tim, I don't see that endpoint in 2022-01-01 redisenterprise.json as well. I'm assuming it is not public?

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (riyasimon23) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json]: Added a new definition Sku Name and changed the two definitions to point to this reference

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (riyasimon23) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:73]: I did notice that I forgot to add /sku endpoint! So I added that recently.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (raosuhas) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1198]: >/ [](http://example.com/codeflow?start=71&length=1)

consider using v3 types

**Review by raosuhas — CHANGES_REQUESTED** [review]
:clock1:

**Review by raosuhas — APPROVED** [review]
:shipit:

**Comment by TimLovellSmith** [comment]
@akning-ms There are some errors for azure powershell and Swagger SDK change that seem spurious to me, could you please check them?

**Comment by TimLovellSmith** [comment]
Actually, now I look closer there seems to be a swagger authoring issue with the SkuName enum.

**Comment by riyasimon23** [comment]
@akning-ms Hi, there are two tests which are failing relating to SDK and Service API. Could you advise me on how to resolve them? I have not yet updated the manifest file yet for this API.

**Comment by akning-ms** [comment]
Hi @riyasimon23  Not sure which SDK failure, if it is for Powershell, you dont have to fix it as it is a optional check. 
for Service API readiness check. it is ensure your service is ready before swagger merge/publish. so when your manifest & service is ready for test, you can trigger pipeline run again by /azp run.  Then ping me to merge if it is passed

**Comment by TimLovellSmith** [comment]
Reopening, we intended to get this merged once it met readiness check but it got automatically closed first.

**Review by TimLovellSmith — CHANGES_REQUESTED** [review]
Waiting for changes from -pr repo to be merged in.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:73]: I think it IS a public endpoint, we just forgot to document it.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1294]: >UnavailableLostCmkAccess [](http://example.com/codeflow?start=9&length=24)

@jantache-microsoft Is this actually going to be something we return any time soon? Today, I wonder a) if we need it. b) whether we might want something more flexible that can return other unavailability reasons.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1850]: >  "readOnly": true, [](http://example.com/codeflow?start=4&length=19)

I'm not sure readonly at this level is meaningful, I thought it only works on properties.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1835]: > "readOnly": true, [](http://example.com/codeflow?start=5&length=18)

I'm not sure readonly at this level is meaningful, I thought it only works on properties.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1889]: >"readOnly": true [](http://example.com/codeflow?start=6&length=16)

I'm not sure readonly at this level is meaningful, I thought it only works on properties.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1901]: It kind of doesn't matter much since its a get only api.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1897]: It kind of doesn't matter much since its a get only api.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1969]: There's a LocationParameter in common types... I see you reference that in one place, but you referenced this in the other, can we just standardize on the common-types one?

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (jantache-microsoft) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1294]: Good point, let's remove it for now, lacking any concrete plans to do extra monitoring around this, and also lacking a concrete plan for using a generic unavailability state value.

Comment (jantache-microsoft) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1294]:
```suggestion
        "Disabled"
```

**Comment by riyasimon23** [comment]
Updated with the suggestions @TimLovellSmith provided.

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:1580]: There's a linter error, not on the diff, that we can fix up by adding x-ms-identifiers notation 

[https://azure.github.io/autorest/extensions/#x-ms-identifiers](https://azure.github.io/autorest/extensions/#x-ms-identifiers)


{"pluginName":"spectral","extensionName":"@microsoft.azure/openapi-validator","level":"error","message":"Missing identifier id in array item property","code":"XmsIdentifierValidation","details":{"jsonpath":["definitions","DatabaseProperties","properties","modules"],"validationCategory":"SDKViolation","providerNamespace":false,"resourceType":false,"range":{"start":{"line":1482,"column":20},"end":{"line":1492,"column":10}}},"source":[{"document":"[file:///mnt/vss/_work/1/lint-c93b354fd9c14905bb574a8834c4d69b/specification/redisenterprise/resource-manager/Microsoft.Cache/stable/2022-01-01/redisenterprise.json](file:///mnt/vss/_work/1/lint-c93b354fd9c14905bb574a8834c4d69b/specification/redisenterprise/resource-manager/Microsoft.Cache/stable/2022-01-01/redisenterprise.json)","position":{"line":1482,"column":9}}]}

File: specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json
Comment (TimLovellSmith) [inline specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:725]: We're getting a linter error which seems to suggest we shouldn't do this, and should instead do

          "final-state-via": "location"

Would that work? Do we actually return a valid Location header?

{"pluginName":"spectral","extensionName":"@microsoft.azure/openapi-validator","level":"error","message":"An async POST operation is tracked via Azure-AsyncOperation header. Set 'final-state-via' property to 'location' on 'x-ms-long-running-operation-options'","code":"PostOperationAsyncResponseValidation","details":{"jsonpath":["paths","/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cache/redisEnterprise/{clusterName}/databases/{databaseName}/import","post"],"validationCategory":"RPaaSViolation","providerNamespace":false,"resourceType":false,"range":{"start":{"line":778,"column":15},"end":{"line":834,"column":8}}},"source":[{"document":"[file:///mnt/vss/_work/1/lint-c93b354fd9c14905bb574a8834c4d69b/specification/redisenterprise/resource-manager/Microsoft.Cache/stable/2022-01-01/redisenterprise.json](file:///mnt/vss/_work/1/lint-c93b354fd9c14905bb574a8834c4d69b/specification/redisenterprise/resource-manager/Microsoft.Cache/stable/2022-01-01/redisenterprise.json)","position":{"line":778,"column":7}}]}

Comment (jantache-microsoft) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:725]: We're not handling the async case for key regeneration right now. So we could make the change to use `location` here:

`          "final-state-via": "location"`

@riyasimon23

Comment (TimLovellSmith) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:725]: @riyasimon23 @jantache-microsoft Yes since the service just always returns 200 today, we should just document that behavior.

Comment (jantache-microsoft) [reply specification/redisenterprise/resource-manager/Microsoft.Cache/preview/2022-11-01-preview/redisenterprise.json:725]: From discussion with Carl and Riya, we decided to go with the change to `"location"`. This avoids any issues around breaking changes / spec issues, in case we want to change to async operation behaviour in the future (since there is an async op in the background, which just isn't exposed right now). It's also the minimal diff to satisfy the linter for now.

**Comment by jantache-microsoft** [comment]
@akning-ms @TimLovellSmith @raosuhas

I'm not sure who's the right person to do the merge - please do another review when you can and merge if it looks good.

Thanks!

**Review by TimLovellSmith — APPROVED** [review]

**Comment by TimLovellSmith** [comment]
@akning-ms ready to merge?

**Pull request merged**
